File: desktop/src/test/java/net/sf/freecol/common/model/IndianSettlementTest.java
package net.sf.freecol.common.model;

import net.sf.freecol.common.model.specification.AbstractGoods;
import net.sf.freecol.common.model.specification.GoodsType;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import promitech.colonization.orders.diplomacy.TradeSession;
import promitech.colonization.savegame.Savegame1600BaseClass;

import static org.assertj.core.api.Assertions.assertThat;

class IndianSettlementTest extends Savegame1600BaseClass {

	GoodsType tradeGoods;
	GoodsType furs;
	Tile nieuwAmsterdamTile;

    @BeforeEach
    public void setup() throws Exception {
		super.setup();
        tradeGoods = goodsType(GoodsType.TRADE_GOODS);
        furs = goodsType(GoodsType.FURS);
        
		nieuwAmsterdamTile = nieuwAmsterdam.tile;
    }
	
	@Test
	public void canGenerateWantedGoods() {
		// given
		Tile tile = game.map.getSafeTile(19, 78);
		
		IndianSettlement is = tile.getSettlement().asIndianSettlement();
		is.getGoodsContainer().decreaseAllToZero();
		
		IndianSettlementWantedGoods sut = new IndianSettlementWantedGoods();
		
		// when
		sut.updateWantedGoods(game.map, is);
		
		// then
		assertThat(is.getWantedGoods()).hasSize(3)
			.extracting(GoodsType::getId)
			.containsAnyOf(
				"model.goods.rum", 
				"model.goods.cigars", 
				"model.goods.cloth", 
				"model.goods.coats" 		
			);
	}
	
	@Test
	public void canGenerateGoodsToSellForWagonTrain() throws Exception {
		// given
		Tile tile = game.map.getSafeTile(19, 78);
		IndianSettlement is = tile.getSettlement().asIndianSettlement();

		Unit wagonTrain = UnitFactory.create(UnitType.WAGON_TRAIN, dutch, nieuwAmsterdamTile);
		wagonTrain.getGoodsContainer().increaseGoodsQuantity(tradeGoods, 100);
		
		IndianSettlementProduction isp = new IndianSettlementProduction();
		isp.init(game.map, is);
		
		// when
		List<? extends AbstractGoods> goodsToSell = isp.goodsToSell(is, wagonTrain);
		
		for (AbstractGoods ag : goodsToSell) {
			System.out.println("goodsToSell " + ag);
		}
		
		// then
		assertThat(goodsToSell)
			.hasSize(3)
			.extracting(AbstractGoods::getQuantity)
			.contains(100, 100, 100);
	}

	@Test
	public void canGenerateGoodsToSellForGalleon() throws Exception {
		// given
		Tile tile = game.map.getSafeTile(19, 78);
		IndianSettlement is = tile.getSettlement().asIndianSettlement();

		Unit galleon = UnitFactory.create(UnitType.GALLEON, dutch, nieuwAmsterdamTile);
		galleon.getGoodsContainer().increaseGoodsQuantity(tradeGoods, 100);
		
		IndianSettlementProduction isp = new IndianSettlementProduction();
		isp.init(game.map, is);
		
		// when
		List<? extends AbstractGoods> goodsToSell = isp.goodsToSell(is, galleon);
		
		for (AbstractGoods ag : goodsToSell) {
			System.out.println("goodsToSell " + ag);
		}
		
		// then
		assertThat(goodsToSell)
			.hasSize(3)
			.extracting(AbstractGoods::getQuantity)
			.contains(25, 25, 25);
	}
	
	@Test
	public void priceForSellGoodsToSettlementWithMissionary() throws Exception {
		// given
		Unit galleon = UnitFactory.create(UnitType.GALLEON, dutch, nieuwAmsterdamTile);
		galleon.getGoodsContainer().increaseGoodsQuantity(tradeGoods, 100);
		Unit wagonTrain = UnitFactory.create(UnitType.WAGON_TRAIN, dutch, nieuwAmsterdamTile);
		wagonTrain.getGoodsContainer().increaseGoodsQuantity(tradeGoods, 100);

		Tile tile = game.map.getSafeTile(19, 78);
		IndianSettlement is = tile.getSettlement().asIndianSettlement();
		is.changeMissionary(UnitFactory.create("model.unit.jesuitMissionary", "model.role.missionary", dutch, is.tile));
		
		TradeSession galleonTradeSession = new TradeSession(game.map, is, galleon)
			.updateSettlementProduction();
		TradeSession wagonTrainTradeSession = new TradeSession(game.map, is, wagonTrain)
			.updateSettlementProduction();
		
		// when
		int galleonSellOfferPrice = galleonTradeSession.sellOffer(tradeGoods, 100);
		int wagonTrainSellOfferPrice = wagonTrainTradeSession.sellOffer(tradeGoods, 100);

		// then
		assertThat(galleonSellOfferPrice).isEqualTo(1270);
		assertThat(wagonTrainSellOfferPrice).isEqualTo(1815);
	}

	@Test
	public void priceForSellGoodsToSettlementWithoutMissionary() throws Exception {
		// given
		Unit galleon = UnitFactory.create(UnitType.GALLEON, dutch, nieuwAmsterdamTile);
		galleon.getGoodsContainer().increaseGoodsQuantity(tradeGoods, 100);
		Unit wagonTrain = UnitFactory.create(UnitType.WAGON_TRAIN, dutch, nieuwAmsterdamTile);
		wagonTrain.getGoodsContainer().increaseGoodsQuantity(tradeGoods, 100);

		Tile tile = game.map.getSafeTile(19, 78);
		IndianSettlement is = tile.getSettlement().asIndianSettlement();
		if (is.hasMissionary()) {
			is.removeMissionary();
		}
		
		TradeSession galleonTradeSession = new TradeSession(game.map, is, galleon)
			.updateSettlementProduction();
		TradeSession wagonTrainTradeSession = new TradeSession(game.map, is, wagonTrain)
			.updateSettlementProduction();
		
		// when
		int galleonSellOfferPrice = galleonTradeSession.sellOffer(tradeGoods, 100);
		int wagonTrainSellOfferPrice = wagonTrainTradeSession.sellOffer(tradeGoods, 100);

		// then
		assertThat(galleonSellOfferPrice).isEqualTo(1050);
		assertThat(wagonTrainSellOfferPrice).isEqualTo(1500);
	}

	@Test
	public void canCalculateBuyPriceFromSettlement() throws Exception {
		// given
		Tile tile = game.map.getSafeTile(19, 78);
		IndianSettlement is = tile.getSettlement().asIndianSettlement();
		is.changeMissionary(UnitFactory.create("model.unit.jesuitMissionary", "model.role.missionary", dutch, is.tile));
		
		Unit galleon = UnitFactory.create(UnitType.GALLEON, dutch, nieuwAmsterdamTile);
		galleon.getGoodsContainer().increaseGoodsQuantity(tradeGoods, 100);
		Unit wagonTrain = UnitFactory.create(UnitType.WAGON_TRAIN, dutch, nieuwAmsterdamTile);
		wagonTrain.getGoodsContainer().increaseGoodsQuantity(tradeGoods, 100);
		
		TradeSession galleonTradeSession = new TradeSession(game.map, is, galleon)
			.updateSettlementProduction();
		TradeSession wagonTrainTradeSession = new TradeSession(game.map, is, wagonTrain)
			.updateSettlementProduction();
		// when
		int galleonBuyPrice = galleonTradeSession.buyOfferPrice(furs, 100);
		int wagonTrainBuyPrice = wagonTrainTradeSession.buyOfferPrice(furs, 100);

		// then
		System.out.println("galleonBuyPrice " + galleonBuyPrice);
		System.out.println("wagonTrainBuyPrice " + wagonTrainBuyPrice);
		
		assertThat(galleonBuyPrice).isEqualTo(116);
		assertThat(wagonTrainBuyPrice).isEqualTo(79);
	}
	
	@Test
    void newTurnProduction() throws Exception {
        // given
        Tile tile = game.map.getSafeTile(19, 78);
        
        IndianSettlement is = tile.getSettlement().asIndianSettlement();

        IndianSettlementProduction isp = new IndianSettlementProduction();

        isp.init(game.map, is);
        
        // when
        isp.updateSettlementGoodsProduction(is);

        // then
        //System.out.println("is.getGoodsContainer() " + is.getGoodsContainer().cloneGoods());
        ProductionSummaryAssert.assertThat(is.getGoodsContainer().cloneGoods())
            .has("model.goods.ore", 1) 
            .has("model.goods.cotton", 200)
            .has("model.goods.furs", 200)
            .has("model.goods.tradeGoods", 16) 
            .has("model.goods.food", 200)
            .has("model.goods.tobacco", 200)
            .has("model.goods.lumber", 76)
        ;
        
    }
	
	
}

File: desktop/src/test/java/net/sf/freecol/common/model/IndianSettlementTest.kt
package net.sf.freecol.common.model

import net.sf.freecol.common.model.IndianSettlementAssert.assertThat
import net.sf.freecol.common.model.ProductionSummaryAssert.assertThat
import net.sf.freecol.common.model.TileAssert.assertThat
import net.sf.freecol.common.model.player.PlayerAssert.assertThat
import net.sf.freecol.common.model.specification.AbstractGoods
import net.sf.freecol.common.model.specification.GoodsType
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import promitech.colonization.orders.diplomacy.TradeSession
import promitech.colonization.savegame.Savegame1600BaseClass

class IndianSettlementTest : Savegame1600BaseClass() {

    lateinit var settlement: IndianSettlement
    lateinit var tradeGoods: GoodsType
    lateinit var furs: GoodsType
    lateinit var nieuwAmsterdamTile: Tile

    @BeforeEach
    override fun setup() {
        super.setup()
        val settlementTile = game.map.getTile(25, 71)
        settlement = settlementTile.settlement.asIndianSettlement()

        nieuwAmsterdamTile = nieuwAmsterdam.tile

        tradeGoods = goodsType(GoodsType.TRADE_GOODS)
        furs = goodsType(GoodsType.FURS)
    }

    @Test
    fun `should increase population`() {
        // given
        settlement.goodsContainer.increaseGoodsQuantity(GoodsType.FOOD, 250)
        val unitsCountBeforeCalculation = settlement.units.size()

        // when
        val indianSettlementProduction = IndianSettlementProduction()
        indianSettlementProduction.init(game.map, settlement)
        indianSettlementProduction.updateSettlementPopulationGrowth(game, settlement)

        // then
        assertThat(settlement).hasUnitsCount(unitsCountBeforeCalculation + 1)
    }

    @Test
    fun `should decrease population when lack of food`() {
        // given
        settlement.goodsContainer.decreaseToZero(GoodsType.FOOD)
        val unitsCountBeforeCalculation = settlement.units.size()

        // when
        val indianSettlementProduction = IndianSettlementProduction()
        indianSettlementProduction.init(game.map, settlement)
        indianSettlementProduction.updateSettlementPopulationGrowth(game, settlement)

        // then
        assertThat(settlement).hasUnitsCount(unitsCountBeforeCalculation - 1)
    }

    @Test
    fun `should destroy settlement when no population and loack of food`() {
        // given
        val settlementTile = settlement.tile
        val settlementOwner = settlement.owner
        settlement.goodsContainer.decreaseToZero(GoodsType.FOOD)
        givenOneUnitInSettlement()

        // when
        val indianSettlementProduction = IndianSettlementProduction()
        indianSettlementProduction.init(game.map, settlement)
        indianSettlementProduction.updateSettlementPopulationGrowth(game, settlement)

        // then
        assertThat(settlement).hasUnitsCount(0)
        assertThat(settlementTile).hasNotSettlement()
        assertThat(settlementOwner).hasNotSettlement(settlement)

    }

    @Test
    fun `should calculate goods production`() {
        // given
        val tile = game.map.getSafeTile(19, 78)
        val settlement = tile.getSettlement().asIndianSettlement()

        val isp = IndianSettlementProduction()

        // when
        isp.init(game.map, settlement)
        isp.updateSettlementGoodsProduction(settlement)

        // then
        //System.out.println("is.getGoodsContainer() " + is.getGoodsContainer().cloneGoods());
        assertThat(settlement.getGoodsContainer().cloneGoods())
            .has("model.goods.ore", 1)
            .has("model.goods.cotton", 200)
            .has("model.goods.furs", 200)
            .has("model.goods.tradeGoods", 16)
            .has("model.goods.food", 200)
            .has("model.goods.tobacco", 200)
            .has("model.goods.lumber", 76)
    }

    @Test
    fun canGenerateWantedGoods() {
        // given
        val tile = game.map.getSafeTile(19, 78)
        val settlement = tile.getSettlement().asIndianSettlement()
        settlement.getGoodsContainer().decreaseAllToZero()

        // when
        IndianSettlementWantedGoods.updateWantedGoods(game.map, settlement)

        // then
        Assertions.assertThat(settlement.getWantedGoods()).hasSize(3)
            .extracting<String, RuntimeException> { obj: GoodsType -> obj.getId() }
            .containsAnyOf(
                "model.goods.rum",
                "model.goods.cigars",
                "model.goods.cloth",
                "model.goods.coats"
            )
    }

    @Test
    fun canGenerateGoodsToSellForWagonTrain() {
        // given
        val tile = game.map.getSafeTile(19, 78)
        val `is` = tile.getSettlement().asIndianSettlement()
        val wagonTrain = UnitFactory.create(UnitType.WAGON_TRAIN, dutch, nieuwAmsterdamTile)
        wagonTrain.goodsContainer.increaseGoodsQuantity(tradeGoods, 100)
        val isp = IndianSettlementProduction()
        isp.init(game.map, `is`)

        // when
        val goodsToSell = isp.goodsToSell(`is`, wagonTrain)
//        for (ag in goodsToSell) {
//            println("goodsToSell $ag")
//        }

        // then
        Assertions.assertThat(goodsToSell)
            .hasSize(3)
            .extracting<Int, java.lang.RuntimeException> { obj: AbstractGoods -> obj.quantity }
            .contains(100, 100, 100)
    }

    @Test
    fun canGenerateGoodsToSellForGalleon() {
        // given
        val tile = game.map.getSafeTile(19, 78)
        val settlement = tile.getSettlement().asIndianSettlement()
        val galleon = UnitFactory.create(UnitType.GALLEON, dutch, nieuwAmsterdamTile)
        galleon.goodsContainer.increaseGoodsQuantity(tradeGoods, 100)
        val isp = IndianSettlementProduction()
        isp.init(game.map, settlement)

        // when
        val goodsToSell = isp.goodsToSell(settlement, galleon)
//        for (ag in goodsToSell) {
//            println("goodsToSell $ag")
//        }

        // then
        Assertions.assertThat(goodsToSell)
            .hasSize(3)
            .extracting<Int, java.lang.RuntimeException> { obj: AbstractGoods -> obj.quantity }
            .contains(25, 25, 25)
    }

    @Test
    fun priceForSellGoodsToSettlementWithMissionary() {
        // given
        val galleon = UnitFactory.create(UnitType.GALLEON, dutch, nieuwAmsterdamTile)
        galleon.goodsContainer.increaseGoodsQuantity(tradeGoods, 100)
        val wagonTrain = UnitFactory.create(UnitType.WAGON_TRAIN, dutch, nieuwAmsterdamTile)
        wagonTrain.goodsContainer.increaseGoodsQuantity(tradeGoods, 100)
        val tile = game.map.getSafeTile(19, 78)
        val settlement = tile.getSettlement().asIndianSettlement()
        settlement.changeMissionary(
            UnitFactory.create("model.unit.jesuitMissionary", "model.role.missionary", dutch, settlement.tile),
            game.turn
        )
        val galleonTradeSession = TradeSession(game.map, settlement, galleon)
            .updateSettlementProduction()
        val wagonTrainTradeSession = TradeSession(game.map, settlement, wagonTrain)
            .updateSettlementProduction()

        // when
        val galleonSellOfferPrice = galleonTradeSession.sellOffer(tradeGoods, 100)
        val wagonTrainSellOfferPrice = wagonTrainTradeSession.sellOffer(tradeGoods, 100)

        // then
        Assertions.assertThat(galleonSellOfferPrice).isEqualTo(1270)
        Assertions.assertThat(wagonTrainSellOfferPrice).isEqualTo(1815)
    }

    @Test
    fun priceForSellGoodsToSettlementWithoutMissionary() {
        // given
        val galleon = UnitFactory.create(UnitType.GALLEON, dutch, nieuwAmsterdamTile)
        galleon.goodsContainer.increaseGoodsQuantity(tradeGoods, 100)
        val wagonTrain = UnitFactory.create(UnitType.WAGON_TRAIN, dutch, nieuwAmsterdamTile)
        wagonTrain.goodsContainer.increaseGoodsQuantity(tradeGoods, 100)
        val tile = game.map.getSafeTile(19, 78)
        val settlement = tile.getSettlement().asIndianSettlement()
        if (settlement.hasMissionary()) {
            settlement.removeMissionary()
        }
        val galleonTradeSession = TradeSession(game.map, settlement, galleon)
            .updateSettlementProduction()
        val wagonTrainTradeSession = TradeSession(game.map, settlement, wagonTrain)
            .updateSettlementProduction()

        // when
        val galleonSellOfferPrice = galleonTradeSession.sellOffer(tradeGoods, 100)
        val wagonTrainSellOfferPrice = wagonTrainTradeSession.sellOffer(tradeGoods, 100)

        // then
        Assertions.assertThat(galleonSellOfferPrice).isEqualTo(1050)
        Assertions.assertThat(wagonTrainSellOfferPrice).isEqualTo(1500)
    }

    @Test
    fun canCalculateBuyPriceFromSettlement() {
        // given
        val tile = game.map.getSafeTile(19, 78)
        val settlement = tile.getSettlement().asIndianSettlement()
        settlement.changeMissionary(
            UnitFactory.create("model.unit.jesuitMissionary", "model.role.missionary", dutch, settlement.tile),
            game.turn
        )
        val galleon = UnitFactory.create(UnitType.GALLEON, dutch, nieuwAmsterdamTile)
        galleon.goodsContainer.increaseGoodsQuantity(tradeGoods, 100)
        val wagonTrain = UnitFactory.create(UnitType.WAGON_TRAIN, dutch, nieuwAmsterdamTile)
        wagonTrain.goodsContainer.increaseGoodsQuantity(tradeGoods, 100)
        val galleonTradeSession = TradeSession(game.map, settlement, galleon)
            .updateSettlementProduction()
        val wagonTrainTradeSession = TradeSession(game.map, settlement, wagonTrain)
            .updateSettlementProduction()
        // when
        val galleonBuyPrice = galleonTradeSession.buyOfferPrice(furs, 100)
        val wagonTrainBuyPrice = wagonTrainTradeSession.buyOfferPrice(furs, 100)

        // then
//        println("galleonBuyPrice $galleonBuyPrice")
//        println("wagonTrainBuyPrice $wagonTrainBuyPrice")
        Assertions.assertThat(galleonBuyPrice).isEqualTo(116)
        Assertions.assertThat(wagonTrainBuyPrice).isEqualTo(79)
    }


    private fun givenOneUnitInSettlement() {
        val first = settlement.units.first()
        for (entity in ArrayList(settlement.units.entities())) {
            settlement.removeUnit(entity)
        }
        settlement.addUnit(first)
    }
}